backend/storage-service/src/jvmMain/kotlin/file/services/SynchronizationService.kt

```diff
@@ -0,0 +1,205 @@
+package dk.sdu.cloud.file.services
+
+import dk.sdu.cloud.Actor
+import dk.sdu.cloud.PageV2
+import dk.sdu.cloud.calls.RPCException
+import dk.sdu.cloud.file.LocalSyncthingDevice
+import dk.sdu.cloud.file.api.*
+import dk.sdu.cloud.file.services.acl.AclService
+import dk.sdu.cloud.file.synchronization.services.SyncthingClient
+import dk.sdu.cloud.service.db.async.*
+import io.ktor.http.*
+import java.io.File
+import java.util.*
+
+object SynchronizedFoldersTable : SQLTable("synchronized_folders") {
+    val id = varchar("id", 36, notNull = true)
+    val device = varchar("device_id", 64, notNull = true)
+    val path = text("path", notNull = true)
+    val accessType = varchar("access_type", 20, notNull = true)
+    val user = text("user_id", notNull = true)
+}
+
+object UserDevicesTable : SQLTable("user_devices") {
+    val device = varchar("device_id", 64, notNull = true)
+    val user = text("user_id", notNull = true)
+}
+
+class SynchronizationService(
+    private val syncthing: SyncthingClient,
+    private val fsPath: String,
+    private val db: DBContext,
+    private val aclService: AclService
+) {
+    private suspend fun chooseFolderDevice(): LocalSyncthingDevice? {
+        return syncthing.config.devices.minByOrNull { device ->
+            db.withSession { session ->
+                session.sendPreparedStatement(
+                    {
+                        setParameter("device", device.id)
+                    },
+                    """
+                        select path
+                        from synchronized_folders
+                        where device_id = :device
+                    """
+                )
+            }.rows.sumOf {
+                CephFsFastDirectoryStats.getRecursiveSize(File(fsPath, it.getField(SynchronizedFoldersTable.path)))
+            }
+        }
+    }
+
+    suspend fun addFolder(actor: Actor, request: SynchronizationAddFolderRequest) {
+        // TODO Check for number of files in folder before adding (limit to 1000_000)
+        if (CephFsFastDirectoryStats.getRecursiveFileCount(File(fsPath, request.path)) > 1000_000) {
+            throw RPCException("Number of files in directory exceeded for synchronization", HttpStatusCode.Forbidden)
+        }
+
+        val id = UUID.randomUUID().toString()
+        val device = chooseFolderDevice()
+        val accessType = if (aclService.hasPermission(request.path, actor.username, AccessRight.WRITE)) {
+            SynchronizationType.SEND_RECEIVE
+        } else {
+            SynchronizationType.SEND_ONLY
```

Comment:
This would allow anyone to read any file in the entire system. You need to check if they have read permissions.